src/sage/schemes/elliptic_curves/hom_sum.py

```diff
@@ -0,0 +1,644 @@
+
+            P,Q = self.domain().change_ring(F).torsion_basis(l**m)
+            if self.is_endomorphism():
+                R,S = P,Q
+            else:
+                R,S = self.codomain().change_ring(F).torsion_basis(l**m)
+            M = self.matrix_on_subgroup((P,Q), (R,S))
+            g = ZZ(gcd(M.list())).p_primary_part(l)
+            if g > 1:
+                scalar *= g
+                M = (M.change_ring(ZZ) / g).change_ring(M.base_ring())
+            K = M.left_kernel_matrix()
+            for row in K:
+                u,v = map(ZZ, row)
+                pt = u*P + v*Q
+                pt.set_order(row.additive_order())
+                ker.append(pt)
+
+        from sage.schemes.elliptic_curves.hom_composite import EllipticCurveHom_composite
+        phi = EllipticCurveHom_composite(self.domain(), [])
+
+        if scalar != 1:
+            phi *= phi.codomain().scalar_multiplication(scalar)
+
+        while ker:
+            K = ker.pop(0)
+
+            (l,e), = K.order().factor()
+            for i in reversed(range(e)):
+                Kl = l**i * K
+                Kl.set_order(l)
+
+                from sage.groups.generic import multiples
+                from sage.misc.misc_c import prod
+                x = polygen(Kl.base_ring())
+                poly = prod(x - T.xy()[0] for T in multiples(Kl, l//2, Kl))
+                poly = poly.change_ring(self.base_ring())
+
+                psi = phi.codomain().isogeny(poly)
+                phi = psi * phi
+                K = psi._eval(K)
+                ker = [psi._eval(P) for P in ker]
+
+        if insep:
+            frob = phi.codomain().frobenius_isogeny(insep)
+            phi = frob * phi
+
+        from sage.schemes.elliptic_curves.hom import find_post_isomorphism
+        iso = find_post_isomorphism(phi, self)
+        return iso * phi
+
+    # EllipticCurveHom methods
+
+    def _degree_bound(self):
+        r"""
+        Return an upper bound on the degree of this sum morphism.
+
+        ALGORITHM: Induction using the bound
+        `|\deg(f+g) - \deg(f) - \deg(g)| \leq 2\sqrt{\deg(f)\cdot\deg(g)}`;
+        see for instance Lemma V.1.2 of [Sil2009]_.
+
+        EXAMPLES::
+
+            sage: E = EllipticCurve(GF(101), [5,5])
+            sage: phi = E.isogenies_prime_degree(7)[0]
+            sage: (phi + phi)._degree_bound()
+            28
+            sage: (phi + phi).degree()
+            28
+            sage: (phi - phi)._degree_bound()
+            28
+            sage: (phi - phi).degree()
+            0
+
+        ::
+
+            sage: E = EllipticCurve(GF(443), [1,1])
+            sage: pi = E.frobenius_endomorphism()
+            sage: m1 = E.scalar_multiplication(1)
+            sage: (pi - m1)._degree_bound()
+            486
+            sage: (pi - m1).degree()
+            433
+        """
+        B = ZZ(0)
+        for phi in self._phis:
+            B += phi.degree() + 2*sqrt(B * phi.degree())
```

Comment:
Is this Cauchy-Schwartz? A comment would be nice. 
Would it not be better to take the integer part each time (within the loop)?  It might give a slightly better bound; more importantly, doing it as written is likely to create some nasty nested square root symbolic expressions.  Probably using isqrt() instead of sqrt) would be a good idea.

Reply:
Thanks, done. I also changed it to compute a lower bound in addition to the upper bound: in many cases the lower bound ends up being trivial anyway, but it could save a little bit of work sometimes.

Reply:
Excellent!